zerver/views/reactions.py

```diff
@@ -66,3 +67,60 @@ def remove_reaction_backend(request, user_profile, message_id, emoji_name):
     do_remove_reaction_legacy(user_profile, message, emoji_name)
 
     return json_success()
+
+@has_request_variables
+def add_reaction(request, user_profile, message_id,
+                 emoji_name=REQ(default=None),
+                 emoji_code=REQ(default=None),
+                 reaction_type=REQ(default="unicode_emoji")):
+    # type: (HttpRequest, UserProfile, int, Text, Text, Text) -> HttpResponse
+    message, user_message = access_message(user_profile, message_id)
+
+    query = Reaction.objects.filter(user_profile=user_profile,
+                                    message=message,
+                                    emoji_code=emoji_code,
+                                    reaction_type=reaction_type)
+    if query.exists():
+        raise JsonableError(_("Reaction already exists."))
+    else:
+        # Reaction doesn't exist so create one.
+        if user_message is None:
+            create_historical_message(user_profile, message)
+
+        query = Reaction.objects.filter(message=message,
+                                        emoji_code=emoji_code,
+                                        reaction_type=reaction_type)
+        if query.exists():
+            # If someone else has reacted to this message with same emoji, use
+            # the emoji name used by them else use the name of user's choice.
+            # Here we should't check for the emoji name consistency as it may be
+            # the case that the reaction is an old one and that emoji may have
+            # been rechristened.
+            emoji_name = query.first().emoji_name
+        else:
+            check_emoji_code_consistency(message.sender.realm, emoji_code, reaction_type)
+            check_emoji_name_consistency(emoji_name, emoji_code, reaction_type)
+        do_add_reaction(user_profile, message, emoji_name, emoji_code, reaction_type)
+
+        return json_success()
+
+@has_request_variables
+def remove_reaction(request, user_profile, message_id,
+                    emoji_code=REQ(default=None),
+                    reaction_type=REQ(default="unicode_emoji")):
+    # type: (HttpRequest, UserProfile, int, Text, Text) -> HttpResponse
+    message, user_message = access_message(user_profile, message_id)
+    query = Reaction.objects.filter(user_profile=user_profile,
+                                    message=message,
+                                    emoji_code=emoji_code,
+                                    reaction_type=reaction_type)
+    if query.exists():
+        # Reaction exists so delete it. While deleting a reaction we don't check
+        # whether the passed data is valid or not since old emoji names and emoji
+        # code may not be consistent with each other according to the current emoji
+        # data. It is safe just delete the reaction if it exists.
+        do_remove_reaction(user_profile, message, emoji_code, reaction_type)
+    else:
+        raise JsonableError(_("Reaction doesn't exist."))
```

Comment:
This would be written better with an early return style, where we do the `JsonableError` first.  I restructured this to do that (and also edited the comment to hopefully be more readable).